pkg/client/client.go

```diff
@@ -1113,6 +1119,12 @@ func (c *Client) CreateOrUpdateSecret(ctx context.Context, s *v1.Secret) error {
 
 	required := s.DeepCopy()
 	mergeMetadata(&required.ObjectMeta, existing.ObjectMeta)
+	if v, ok := existing.Data["tls.crt"]; ok && len(v) > 0 {
```

Comment:
shouldn't we check if the service CA annotation is present and if yes, assume that the `Data` field is fully managed by the service CA operator (e.g. CMO only manages the metadata)?

Reply:
Checking the annotation is a good point yes.

While it's true for CMO, that only the service-ca-operator will touch the data field, simply by convention, I wouldn't put this in the code. The service-ca-operator will only update the keys that are checked here. Other potentially present keys would not be altered.
The current code will already remove data keys added by users. Should we ever add any fields to any Secret manifest, we might produce a subtle bug by copying the whole data field here.

Reply:
oh and there won't be an annotation on the Secret that we can check. But an annotation check on ConfigMaps and Vaildatingwebhooks makes sense.

Reply:
On a Secret several fields will be added by servce-ca-operator, though not sure if checking any of those makes sense:
- service-ca-operator will add a `ManagedField` like so:
```
managedFields:
    - manager: service-ca-operator
      ...
  ```
  - A type field is added `type: kubernetes.io/tls`
  - An annotation with the expiry date is added `annotations: service.beta.openshift.io/expiry: '2023-11-29T06:22:59Z'`
  
  All these would work, not sure if we should check on of those though.

Reply:
Right I confused the annotation that CMO puts on the service with the resulting secret... 
From what I can tell:
* For any configmap with the `config.openshift.io/inject-trusted-cabundle=true` label ([custom CA bundle](https://docs.openshift.com/container-platform/4.9/networking/configuring-a-custom-pki.html#certificate-injection-using-operators_configuring-a-custom-pki)) or `service.beta.openshift.io/inject-cabundle=true` annotation ([service CA bundle](https://docs.openshift.com/container-platform/4.9/security/certificates/service-serving-certificate.html#add-service-certificate-configmap_service-serving-certificate)), CMO should keep the Data field untouched.
* The same goes for validating webhooks and API services with the `service.beta.openshift.io/inject-cabundle=true` annotation.
* For secrets, the situation is a bit tricky... Maybe CMO can assume that a secret of type `kubernetes.io/tls` with at least one annotation starting with the `service.beta.openshift.io/` prefix is managed by the service CA operator and the data shouldn't be modified?

Reply:
I wonder if this is the right place to make the change because now `CreateOrUpdateSecret` needs domain specific knowledge about managing secret contents. Instead, I would expect the caller to manage the contents of the secret, and `CreateOrUpdateSecret` should just persist whatever is provided by the caller. 

Also, when it comes to TLS secrets, shouldn't CMO just create these secrets and never ever update them?

Reply:
The various `CreateOrUpdate` functions in CMO already do this today. Since this is not a generic interface I don't see this as an issue.

Regarding the TLS secrets, CMO today follows the paradigm that all resources are immutable. As this PR shows allowances need to be made, but I don't see a reason to not keep reverting unwanted changes (say labels a user might attach to a TLS secret). I'm also unclear how we would reconcile changes that we want CMO to deploy on a version update.

Reply:
> Also, when it comes to TLS secrets, shouldn't CMO just create these secrets and never ever update them?

I think that CMO still needs to update the secrets because CMO may add/remove metadata (labels, annotations, owner references) between versions ([example](https://github.com/openshift/cluster-monitoring-operator/commit/0bc17b108cac36b09c10c355e0533101f9efdd3f#diff-0855bf7fb0e348899b2128ea27ba8725f294c9d7a28609925de4ccd1f94667b0)).